Remove ParametrizedSchedule dependency from instmap

nkanazawa1989 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
close #5333 

### Details and comments
this is draft PR.

TODO:
- [x] address below issues
- [x] update unittest
- [x] add reno

### 1. Storing `Callable` in instmap

Basically we don't need to allow instmap to take callback function to parse backend defaults. However, in some edge cases, we may want to allow this.

Let's think about the situation we want to choose the pulse shape based on some parameter.
```
parametrize width.
if width is less than zero 0:
   create the schedule with Gaussian
else:
   create the schedule with GaussianSquare
```
More specifically, the question is equivalent to _if we want to create schedule with control statement_. If we don't need to think about this situation, we just need to deal with `Schedule` and the logic of `InstructionScheduleMap.get` and `InstructionScheduleMap.add` can be simplified. Currently program in instmap is `Union[Schedule, Callable]`.

Though we don't allow `Callable` in instmap, we are still able to implement the program in above example with pulse factory that returns Gaussian or GaussianSquare depending on width. The class may look like
```python
class MetaGaussian:
    def __new__(cls, *args):
        if args[3] > 0:
            return pulse.GaussianSquare(*args)
        else:
            return pulse.Gaussian(*args[:-1])
```

### 2. Handling of identical parameters in U* schedules

In current IBMQ backend, U* gates have multiple frame change parameters across channels to synchronize them. For example, u2 may consist of
```
[PulseQobjInstruction(name="parametric_pulse", t0=0, ch="d0", pulse_shape="drag", parameters={'amp': (0.0002822041705496269+0.09803512541692665j), 'beta': 0.811851744060955, 'duration': 160, 'sigma': 40}),
 PulseQobjInstruction(name="fc", t0=0, ch="d0", phase="-(P1)"),
 PulseQobjInstruction(name="fc", t0=160, ch="d0", phase="-(P0)"),
 PulseQobjInstruction(name="fc", t0=0, ch="u1", phase="-(P1)"),
 PulseQobjInstruction(name="fc", t0=160, ch="u1", phase="-(P0)")]
```

If we simply create parametrized schedule from operand name, ie
```python
phase = circuit.Parameter(inst.phase)
```
we way generate independent parameter objects with the same name, i.e. `{Parameter(P0), Parameter(P0), Parameter(P1), Parameter(P1)}`. Actually both `P0` (`P1`) objects refer to the same parameter `'P0'` (`'P1'`), but circuit `Parameter` doesn't allow us to define parameter by name (`Parameter` is implicitly defined by uuid). Due to this limitation, we need to insert ugly logic to manage mapping of parameter name and `Parameter` object.

https://github.com/nkanazawa1989/qiskit-terra/blob/7813dddf9d980e41b26980419223c5ecae197baa/qiskit/qobj/converters/pulse_instruction.py#L441-L469.

I don't think this implementation is right direction.

eggerdj commented on qiskit/qobj/converters/pulse_instruction.py:
```diff
@@ -553,19 +564,7 @@ def convert_set_frequency(self, instruction):
-            return ParameterizedSchedule(gen_sf_schedule, parameters=frequency_expr.params)
-        else:
-            frequency = frequency * GIGAHERTZ_TO_SI_UNITS
+        frequency = self.format_value(instruction.phase) * GIGAHERTZ_TO_SI_UNITS
```
Should this not be `instruction.frequency`?

eggerdj commented on qiskit/qobj/converters/pulse_instruction.py:
```diff
@@ -583,19 +582,7 @@ def convert_shift_frequency(self, instruction):
-            return ParameterizedSchedule(gen_sf_schedule, parameters=frequency_expr.params)
-        else:
-            frequency = frequency * GIGAHERTZ_TO_SI_UNITS
+        frequency = self.format_value(instruction.phase) * GIGAHERTZ_TO_SI_UNITS
```
```suggestion
        frequency = self.format_value(instruction.frequency) * GIGAHERTZ_TO_SI_UNITS
```

eggerdj reviewed (changes requested):
Are there some tests that you can add to test the behavior?

eggerdj commented:
Perhaps we do not need to store a callable in inst map. Maybe the functionality of having the type of pulse shape depend on a parameter can be implemented with child classes of ParametricPulse?

nkanazawa1989 commented:
Thanks @eggerdj for reviewing and comment! Currently I have several uncertain (temporarily) implementations regarding to two issues and skip writing tests until I find proper implementation. In my local environment this code can generate `Schedule`-based instmap from backend object, just to be sure.

> Maybe the functionality of having the type of pulse shape depend on a parameter can be implemented with child classes of ParametricPulse?

Yes. This can allow us to eliminate Callable from instmap. For point 1 I also think this approach is reasonable and want to remove Callable.

lcapelluto commented on qiskit/pulse/instruction_schedule_map.py:
```diff
@@ -164,11 +165,36 @@ def get(self,
-        # otherwise this is just a Schedule
+
+        # schedule
+        sched_params = list(schedule_generator.parameters)
```
these have no natural ordering

lcapelluto replied:
This might change, see linked PR

lcapelluto commented:
I have to keep thinking about issue 1.

As far as issue 2, I don't think the implementation of `format_value` is so bad. The additions to `inst_map` do look rather complicated, but I don't think it's the wrong approach. We could add some methods to `Schedule`, for instance, `get_parameter(self, parameter_name)`, to simplify a bit of the logic.

There might be a bit of an issue right now with ordering. `inst_map` allows params and yet there's no real ordering with the actual parameters. Here is what I mean:

```
 [PulseQobjInstruction(name="fc", t0=0, ch="d0", phase="-(P2*P1)")]
```
Even with only one `ParameterExpression`, there is no ordering of the parameters `P1` and `P2`, so this trickles up to `Schedule.parameters` as well.

If we only accept `kwargs` with `inst_map`, that would be the best solution. Otherwise, we might need:
`p1 = Parameter("P1")`
`p2 = Parameter("P2")`
`Parameter("index_0") -> p1`
`Parameter("index_1") -> p2`
`Parameter("index_3") -> p2`
`Parameter("index_4") -> p1`
That would be even more complicated, and I'm not sure it would actually work. I'm not sure that positional arguments will be easy to support everywhere -- but I'm pretty sure they can't be supported by only a subset of the stack. You are more familiar with this problem than I am. How important is it to support positional schedule parameters? Would we be able to phase them out?

lcapelluto commented:
https://github.com/Qiskit/qiskit-terra/pull/3910 Actually, there is some discussion about making circuit parameters ordered

nkanazawa1989 commented:
Thanks @lcapelluto for insightful comments. 

For 1 I keep Callback as is. For 2 as you suggested the ordering of parameter is the serious issue which I was missing, but this is addressed in latest commits. The signature of `InstructionScheduleMap.add` becomes `add(self, instructions, qubits, schedule, params: Optional[List[str]])` here I added `params` to manage parameter order of given instruction. For example, if we define parameters `theta, phi, lambda` in this order, we can specify `params=['theta', 'phi', 'lambda']` so that the instmap can keeps parameter ordering when `.get` method is called. This is optional argument to keep backward compatibility, and the parameter order defaults to the alphanumerical ordering. Because each instmap element has name based parameter definition, we can merge individual `Parameter` objects with the same name - thus we don't need `format_value` any more.

I also added signature check based on this new information.
https://github.com/Qiskit/qiskit-terra/blob/292a2fe541ab4fd15fe9210b827e2683aef7ff13/qiskit/pulse/instruction_schedule_map.py#L181-L185

I believe this behavior is very common in standard python function. Previously this check was just ignored and we were able to assign arbitrary number of parameters to instructions even if they are not parameterized - this is actually a defect which could induce a bug.

However this check caused many troubles in unittests that have dependency on `FakeOpenPulse2Q` or `FakeOpenPulse3Q`.

https://github.com/Qiskit/qiskit-terra/blob/9e6a016dbd44fdeedf27241b9e02fbd8d4a617c8/qiskit/test/mock/fake_openpulse_2q.py#L185-L191

As you can see in the definition, `u3` instruction has no parameter though U3 gate cannot be defined without parameter. Now tests for measurement alignment and some part of pulse builder failed because they try to assign three parameters with implicit assumption of `u3(P0, P1, P2)`. This results in `PulseError` because instruction has no parameter.

I think the design of `FakeBackend` is defect and need to fix this rather than removing signature check. However many modules may have dependency on fake backend, thus I think I cannot simply overwrite this configuration (i.e. pulse simulator may use this?). I think I have three options to step forward.
- Define `FakeOpenPulse2Qv2` backend and replace backend of specific unittests causing the trouble.
- Overwrite existing `FakeOpenPulse2Q`.
- Remove signature check.

nkanazawa1989 commented:
Misc changes for code review

- Added format function to `pulse.instructions.Instruction`
In this PR `ParameterizedSchedule` is replaced with `Schedule` and operands of an instruction become `ParameterExpression`. The assigned `ParameterExpression` object and a reference `float` value cannot be compared and thus equality check always fails. All `ParameterExpression` should be converted into suitable data type.

https://github.com/nkanazawa1989/qiskit-terra/blob/292a2fe541ab4fd15fe9210b827e2683aef7ff13/qiskit/pulse/instructions/instruction.py#L380-L381

- Partial parameter binding is allowed for the instmap `.get` method.
Previously partial parameter binding was not allowed because `ParameterizedSchedule` cannot be a part of `Schedule`. In this PR, unassigned schedules are also `Schedule` thus we can allow for partial binding. Some unittests requiring all parameter binding are removed accordingly.

lcapelluto commented on qiskit/pulse/instruction_schedule_map.py:
```diff
@@ -161,39 +169,89 @@ def get(self,
-        schedule_generator = self._map[instruction].get(_to_tuple(qubits))
+        schedule_param_tuple = deepcopy(self._map[instruction][_to_tuple(qubits)])
+
+        # create parameter-value mapping
+        if len(params) > len(schedule_param_tuple.parameters):
+            raise PulseError('Too many values to bind: {}.'.format(', '.join(map(str, params))))
+
+        if not all(key in schedule_param_tuple.parameters for key in kwparams.keys()):
```
```suggestion
        # Verify parameter-value mapping
        if len(params) > len(schedule_param_tuple.parameters):
            raise PulseError('Too many values to bind: {}.'.format(', '.join(map(str, params))))
        if not all(key in schedule_param_tuple.parameters for key in kwparams.keys()):
```

lcapelluto commented on qiskit/pulse/instruction_schedule_map.py:
```diff
@@ -161,39 +169,89 @@ def get(self,
+            return schedule_param_tuple.schedule(**bind_parameters)
+
+        # schedule
+        sched_params = list(schedule_param_tuple.schedule.parameters)
+        if sched_params:
```
```suggestion
        sched = schedule_param_tuple.schedule
        if sched.is_parameterized():
```

lcapelluto replied:
Recommending this pair of changes to make it more clear that we are not looking at `schedule_param_tuple.parameters`

lcapelluto commented on qiskit/pulse/instruction_schedule_map.py:
```diff
@@ -161,39 +169,89 @@ def get(self,
+        sched_params = list(schedule_param_tuple.schedule.parameters)
+        if sched_params:
+            parameter_mapping = dict()
+            for sched_param in sched_params:
```
```suggestion
            for sched_param in sched.parameters:
```

lcapelluto commented on qiskit/pulse/instruction_schedule_map.py:
```diff
@@ -161,39 +169,89 @@ def get(self,
         instruction = _get_instruction_string(instruction)
         self.assert_has(instruction, qubits)
-        schedule_generator = self._map[instruction].get(_to_tuple(qubits))
+        schedule_param_tuple = deepcopy(self._map[instruction][_to_tuple(qubits)])
```
nitpick: It is a little challenging keeping `schedule.parameters`, `params`, and `schedule_param_tuple.parameters` straight.

```
schedule  # this is clear
schedule.parameters   # this seems clear
params     # input_params? as an alias within the function?
kwparams  # seems clear that its keyword args
schedule_param_tuple.parameters = schedule_arguments  # ?? I'm not sure
```

lcapelluto commented on qiskit/pulse/instruction_schedule_map.py:
```diff
@@ -161,39 +169,89 @@ def get(self,
+            func_parameters = set(param.name for param in schedule.parameters)
+
+        if params is None:
+            if len(func_parameters) > 0:
+                # for backward compatibility
+                params = sorted(func_parameters)
+            else:
+                params = []
```
`sorted([])` works fine
```suggestion
            # for backward compatibility
            params = sorted(func_parameters)
```

lcapelluto reviewed (changes requested):
This is awesome, Naoki!

lcapelluto commented:
Regarding the fake backends, I agree that they are incorrect. I'd say they should be updated to be more correct, except when used to test bad backend input. Hopefully other repos don't depend on the incorrect mock backends -- strictly speaking, they should be for terra tests, but I know they're useful for other use cases sometimes, too. Have you tried out this approach yet? We could ask @ajavadia and @chriseclectic if there might be unexpected dependencies :)

nkanazawa1989 commented:
Thanks Lauren! I updated both FakeOpenPulse2Q and 3Q along with all associated unittest references. I've carefully read those test codes and I think I made correct updates, but please double check these unittest updates and new backend models. I think this PR is ready for the final review :)

lcapelluto commented on qiskit/pulse/instructions/instruction.py:
```diff
@@ -109,7 +110,7 @@ def id(self) -> int:  # pylint: disable=invalid-name
     def operands(self) -> Tuple:
         """Return instruction operands."""
-        return self._operands
+        return tuple(map(_format_value, self._operands))
```
Could this be called earlier, for performance? I believe operands can only be set upon initialization and updated via `assign_parameters`

nkanazawa1989 replied:
I think think we need to format this on the fly because in the calibration module we directly create parameterized schedule though it's not necessary to compare two schedules so far. If `assign_parameters` is immutable, we can take that approach. For performance point of view maybe we can use [LRU cache](https://docs.python.org/3/library/functools.html#functools.lru_cache).

lcapelluto commented on qiskit/pulse/schedule.py:
```diff
@@ -716,6 +717,21 @@ def assign_parameters(self,
 
         return self
 
+    def get_parameter(self,
+                      parameter_name: str) -> List[Parameter]:
```
```suggestion
    def get_parameters(self,
                       parameter_name: str) -> List[Parameter]:
```
I think this should be plural, since it returns a list.

lcapelluto reviewed (commented):
I only had a couple nitpicks. This PR is awesome! 🎉 

Are there any tests which use and demo the new `arguments` option in InstructionScheduleMap? Something where the default alphanumeric ordering isn't used.

lcapelluto commented on qiskit/pulse/parser.py:
```diff
@@ -182,10 +182,23 @@ def visit_Num(self, node: ast.Num) -> ast.Num:
 
-    def visit_Name(self, node: ast.Name) -> Union[ast.Name, ast.Num]:
-        """Evaluate name and return ast.Num if it is bound.
+    def visit_Constant(self, node: ast.Constant) -> ast.Constant:
```
Just for my own understanding, where did this come up?

nkanazawa1989 replied:
Please check this https://docs.python.org/3/library/ast.html#ast.AST

> Deprecated since version 3.8: Old classes ast.Num, ast.Str, ast.Bytes, ast.NameConstant and ast.Ellipsis are still available, but they will be removed in future Python releases. In the meantime, instantiating them will return an instance of a different class.

Now constructor of `ast.Num` returns `ast.Constant` and the parser checks if all parameters are bound by calling `isinstance` (previously ast node becomes `ast.Num` is expression is evaluated but now `ast.Constant` and this check caused bug).

lcapelluto commented:
Looking forward to this!

nkanazawa1989 commented on qiskit/pulse/instructions/instruction.py:
```diff
@@ -110,7 +111,8 @@ def id(self) -> int:  # pylint: disable=invalid-name
     def operands(self) -> Tuple:
         """Return instruction operands."""
-        return tuple(map(_format_value, self._operands))
+        return tuple([_format_value(operand) if isinstance(operand, ParameterExpression)
+                      else operand for operand in self._operands])
```
I think so

nkanazawa1989 commented:
Finally all checks have passed. I also modified parameter formatting mechanism (previously this is formatted at each call but now format is applied only once when parameters are assigned -- thus smaller overhead).

nkanazawa1989 commented on qiskit/pulse/instructions/instruction.py:
```diff
@@ -275,7 +276,8 @@ def assign_parameters(self,
             for op_idx in op_indices:
-                new_operands[op_idx] = new_operands[op_idx].assign(parameter, value)
+                param_expr = new_operands[op_idx]
+                new_operands[op_idx] = format_parameter_value(param_expr.assign(parameter, value))
```
Though all test have passed, this could cause problem in parameterization of channel index because it seems like evaluating timeslot update based on value type (in Schedule timeslot is updated if channel index value is ParameterExpression but this format value before the evaluation).

lcapelluto replied:
I don't quite follow. Could you write a test that would cause this to break?

nkanazawa1989 replied:
I carefully read the code and concluded this is not a problem. Because channel index is parameterized at schedule level, so converting ParameterExpression into formatted type at instruction level doesn't affect channel parameterization (channel index is still ParameterExpression).

lcapelluto commented on qiskit/pulse/library/parametric_pulses.py:
```diff
@@ -122,13 +124,7 @@ def assign_parameters(self,
-                    except TypeError:
-                        # It's alright if the value is still parameterized
-                        pass
+                    op_value = format_parameter_value(op_value.assign(parameter, value))
```
nice 👍

eggerdj reviewed (approved):
LGTM
